README.md
# promptpay

This is for generate promptpay QR data.

## support platform
- e-wallet
- identity/tax-id
- phone number

CHANGELOG.md
## [1.4.0] - 2020 March 18
### Features
- feat: add isQRDataValid() to check is this QR PromptPay

## [1.3.1] - 2020 March 18
- format file with `flutter format`

## [1.3.0] - 2020 March 16
### Features
- feat: can change amount of existing qr code
### Fix
- fix: get account number can extract merchant id

## [1.2.1] - 2020 March 13
- fix: wrong typo in example

## [1.2.0] - 2020 March 13
- add getAccountNumberFromQRData(String qrData)

## [1.1.0] - 2020 Febuary 27
- change amount to optional argument

## [1.0.2] - 2020 Febuary 26
- change homepage to github

## [1.0.1] - 2020 Febuary 26
- standardize project

## [1.0.0] - 2020 Febuary 26

- support three account types - e-wallet | identity/tax-id | phone number


test/promptpay_test.dart
import 'dart:convert';

import 'package:either_option/either_option.dart';
import 'package:flutter_test/flutter_test.dart';

import 'dart:core';
import 'package:promptpay/promptpay.dart';
import 'package:promptpay/promptpay_data.dart';

void main() {
  test('toString 3 must be 03', () {
    expect(3.toString().padLeft(2, '0'), "03");
  });
  test('generate qr data without amount', () {
    expect(PromptPay.generateQRData("0000000000"), "00020101021129370016A000000677010111011300660000000005802TH530376463048956");
  });
  test('should generate promptpay data for phone number correctly', () {
    expect(PromptPay.generateQRData("0812345678", amount: 123.45), "00020101021129370016A000000677010111011300668123456785802TH53037645406123.4563043045");
  });
  test('should generate promptpay data for identity correctly', () {
    expect(PromptPay.generateQRData("1100599000373", amount: 123.45), "00020101021129370016A000000677010111021300000000000005802TH53037645406123.4563046DCC");
  });

  test('should generate promptpay data for e-wallet correctly', () {
    expect(PromptPay.generateQRData("000000000000000", amount: 123.45), "00020101021129370016A00000067701011103150000000000000005802TH53037645406123.456304AE16");
  });

  test('qr data should be valid', () {
    final qrData = "00020101021129370016A000000677010111011300668123456785802TH53037645406123.4563043045";
    expect(PromptPay.isQRDataValid(qrData), true);
  });

  test('it is not qr data should not be valid', () {
    final notQRData = "31902380129830197401412890841209";
    expect(PromptPay.isQRDataValid(notQRData), false);
  });

  test('it is not qr data because length < 8', () {
    final notQRData = "1231356";
    expect(PromptPay.isQRDataValid(notQRData), false);
  });

  test('extract qrcode from qr data', () {
    final qrData = "00020101021129370016A000000677010111011300668123456785802TH53037645406123.4563043045";
    final indexOfStartApplicationData = qrData.indexOf(applicationIDData);
    
    expect(indexOfStartApplicationData, 20);
    expect(qrData[indexOfStartApplicationData], "A");

    final indexOfStartAccountType = indexOfStartApplicationData + applicationIDData.length;
    
    expect(indexOfStartAccountType, 36);

    final accountTypeString = qrData.substring(indexOfStartAccountType, indexOfStartAccountType + 2);

    expect(accountTypeString, "01");

    final accountType = accountTypeString == "01"
      ? AccountType.phone : accountTypeString == "02"
      ? AccountType.identityNumber : AccountType.eWallet;

    expect(accountType, AccountType.phone);

    final indexOfStartAccountLength = indexOfStartAccountType + 2;
    final accountLength = int.parse(qrData.substring(indexOfStartAccountLength, indexOfStartAccountLength + 2));

    expect(accountLength, 13);

    final indexOfStartAccountData = indexOfStartAccountLength + 2;
    final accountData = qrData.substring(indexOfStartAccountData, indexOfStartAccountData + accountLength);

    expect(accountData, "0066812345678");
  });

  test('extract qrcode from qr data with function', () {
    final qrData = "00020101021129370016A000000677010111011300668123456785802TH53037645406123.4563043045";
    expect(PromptPay.getAccountNumberFromQRData(qrData), Option.of("0066812345678"));
  });

  test('test promptpay constructor', () {
    final qrData = "00020101021129370016A000000677010111011300668123456785802TH53037645406123.4563043045";
    var promptPayData = PromptPayData.fromQRData(qrData);
    expect(promptPayData.emvcoVersion.type, PromptPayFieldType.EMVCoVersion);
    expect(promptPayData.emvcoVersion.length, 2);
    expect(promptPayData.emvcoVersion.data, "01");
    
    expect(promptPayData.qrType.type, PromptPayFieldType.qrType);
    expect(promptPayData.qrType.length, 2);
    expect(promptPayData.qrType.data, "11");

    expect(promptPayData.transferring.type, PromptPayFieldType.transferring);
    expect(promptPayData.transferring.length, 37);
    expect(promptPayData.transferring.data, "0016A00000067701011101130066812345678");
  });
}

lib/promptpay_data.dart
enum PromptPayFieldType {
  unknown,
  EMVCoVersion,
  qrType,
  transferring,
  transferringType,
  transferringPhoneNumber,
  transferringIdentityNumber,
  transferringEWallet,
  billing,
  billingType,
  billingID,
  billingReferenceNumber1,
  billingReferenceNumber2,
  country,
  currency,
  amount,
  unknownFieldTypeForBilling,
  checksum
}

class PromptPayField {
  static const _mapIDNumberWithField = {
    "00": PromptPayFieldType.EMVCoVersion,
    "01": PromptPayFieldType.qrType,
    "29": PromptPayFieldType.transferring,
    "29_00": PromptPayFieldType.transferringType,
    "29_01": PromptPayFieldType.transferringPhoneNumber,
    "29_02": PromptPayFieldType.transferringIdentityNumber,
    "29_03": PromptPayFieldType.transferringEWallet,
    "30": PromptPayFieldType.billing,
    "30_00": PromptPayFieldType.billingType,
    "30_01": PromptPayFieldType.billingID,
    "30_02": PromptPayFieldType.billingReferenceNumber1,
    "30_03": PromptPayFieldType.billingReferenceNumber2,
    "53": PromptPayFieldType.currency,
    "54": PromptPayFieldType.amount,
    "58": PromptPayFieldType.country,
    "62": PromptPayFieldType.unknownFieldTypeForBilling,
    "63": PromptPayFieldType.checksum,
  };

  PromptPayFieldType type;
  String typeID;
  int length;
  String data;

  PromptPayField();

  PromptPayField.fromSubStringOfQRData(String fieldQRData) {
    type = determineFieldType(fieldQRData.substring(0, 2));
    typeID = fieldQRData.substring(0, 2);
    length = int.parse(fieldQRData.substring(2, 4));
    data = fieldQRData.substring(4, 4 + length);
  }

  PromptPayField.fromTransferringOrBillingSubType(String fieldQRData) {
    type = determineFieldType(fieldQRData.substring(0, 5));
    typeID = fieldQRData.substring(3, 5);
    length = int.parse(fieldQRData.substring(5, 7));
    data = fieldQRData.substring(7, 7 + length);
  }

  static PromptPayFieldType determineFieldType(String fieldData) {
    if (_mapIDNumberWithField[fieldData] == null) {
      return PromptPayFieldType.unknown;
    }

    return _mapIDNumberWithField[fieldData];
  }
}

class PromptPayData {
  PromptPayField emvcoVersion;
  PromptPayField qrType;
  PromptPayField transferring;
  PromptPayField transferringSubType;
  PromptPayField transferringPhoneNumber;
  PromptPayField transferringIdentityNumber;
  PromptPayField transferringEWallet;
  PromptPayField billing;
  PromptPayField billingType;
  PromptPayField billingID;
  PromptPayField billingReferenceNumber1;
  PromptPayField billingReferenceNumber2;
  PromptPayField country;
  PromptPayField currency;
  PromptPayField amount;
  PromptPayField unknownFieldTypeForBilling;
  PromptPayField checksum;

  PromptPayData.fromQRData(String qrData) {
    var currentIndex = 0;
    while (currentIndex < qrData.length) {
      var dataLength =
          int.parse(qrData.substring(currentIndex + 2, currentIndex + 4));
      var promptPayField = PromptPayField.fromSubStringOfQRData(
          qrData.substring(currentIndex, currentIndex + 2 + 2 + dataLength));

      switch (promptPayField.type) {
        case PromptPayFieldType.unknown:
          break;
        case PromptPayFieldType.EMVCoVersion:
          emvcoVersion = promptPayField;
          break;
        case PromptPayFieldType.qrType:
          qrType = promptPayField;
          break;
        case PromptPayFieldType.transferring:
          transferring = promptPayField;
          break;
        case PromptPayFieldType.billing:
          billing = promptPayField;
          break;
        case PromptPayFieldType.country:
          country = promptPayField;
          break;
        case PromptPayFieldType.currency:
          currency = promptPayField;
          break;
        case PromptPayFieldType.amount:
          amount = promptPayField;
          break;
        case PromptPayFieldType.unknownFieldTypeForBilling:
          unknownFieldTypeForBilling = promptPayField;
          break;
        case PromptPayFieldType.checksum:
          checksum = promptPayField;
          break;
        default:
          break;
      }

      currentIndex = currentIndex + 2 + 2 + dataLength;
    }

    // extract subtype
    if (transferring != null) {
      var trasferringCurrentIndex = 0;
      while (trasferringCurrentIndex < transferring.data.length) {
        var dataLength = int.parse(transferring.data.substring(
            trasferringCurrentIndex + 2, trasferringCurrentIndex + 4));
        var transferringPromptPayField =
            PromptPayField.fromTransferringOrBillingSubType(
                "29_${transferring.data.substring(trasferringCurrentIndex, trasferringCurrentIndex + 2 + 2 + dataLength)}");

        switch (transferringPromptPayField.type) {
          case PromptPayFieldType.transferringType:
            transferringSubType = transferringPromptPayField;
            break;
          case PromptPayFieldType.transferringPhoneNumber:
            transferringPhoneNumber = transferringPromptPayField;
            break;
          case PromptPayFieldType.transferringIdentityNumber:
            transferringIdentityNumber = transferringPromptPayField;
            break;
          case PromptPayFieldType.transferringEWallet:
            transferringEWallet = transferringPromptPayField;
            break;
          default:
            break;
        }

        trasferringCurrentIndex = trasferringCurrentIndex + 2 + 2 + dataLength;
      }
    }

    if (billing != null) {
      var billingCurrentIndex = 0;
      while (billingCurrentIndex < billing.data.length) {
        var dataLength = int.parse(billing.data
            .substring(billingCurrentIndex + 2, billingCurrentIndex + 4));
        var billingPromptPayField = PromptPayField.fromTransferringOrBillingSubType(
            "30_${billing.data.substring(billingCurrentIndex, billingCurrentIndex + 2 + 2 + dataLength)}");

        switch (billingPromptPayField.type) {
          case PromptPayFieldType.billingType:
            billingType = billingPromptPayField;
            break;
          case PromptPayFieldType.billingID:
            billingID = billingPromptPayField;
            break;
          case PromptPayFieldType.billingReferenceNumber1:
            billingReferenceNumber1 = billingPromptPayField;
            break;
          case PromptPayFieldType.billingReferenceNumber2:
            billingReferenceNumber2 = billingPromptPayField;
            break;
          default:
            break;
        }

        billingCurrentIndex = billingCurrentIndex + 2 + 2 + dataLength;
      }
    }
  }

  Iterable<PromptPayField> asIterable() {
    return [
      emvcoVersion,
      qrType,
      transferring,
      billing,
      country,
      currency,
      amount,
      unknownFieldTypeForBilling
    ];
  }
}


lib/promptpay.dart
library promptpay;

import 'dart:convert';
import 'dart:core';

import 'package:crclib/crclib.dart';
import 'package:either_option/either_option.dart';
import 'package:promptpay/promptpay_data.dart';

const versionID = "00";
const qrTypeID = "01";
const merchantAccountID = "29";
const subMerchantApplicationID = "00";
const subMerchantAccountPhoneID = "01";
const subMerchantAccountIdentityID = "02";
const subMerchantAccountEWalletID = "03";
const countryID = "58";
const currencyID = "53";
const amountID = "54";
const checksumID = "63";

const versionLength = "02";
const qrTypeLength = "02";
const merchantAccountLength = "37";
const subMerchantApplicationIDLength = "16";
const subMerchantAccountLength = "13";
const countryLength = "02";
const currencyLength = "03";
const checksumLength = "04";

const versionData = "01";
const qrMultipleTypeData = "11";
const qrOneTimeTypeData = "12";
const applicationIDData = "A000000677010111";
const countryData = "TH";
const bahtCurrencyData = "764";

enum AccountType { phone, identityNumber, eWallet }

/// A PromptPay is a payment method in Thailand
class PromptPay {
  /// Returns [QR Code Data] for PromptPay QR code
  static String generateQRData(String target, {double amount}) {
    AccountType accountType = target.length >= 15
        ? (AccountType.eWallet)
        : target.length >= 13
            ? (AccountType.identityNumber)
            : (AccountType.phone);

    var data = [
      versionID,
      versionLength,
      versionData,
      qrTypeID,
      qrTypeLength,
      qrMultipleTypeData,
      merchantAccountID,
      merchantAccountLength,
      subMerchantApplicationID,
      subMerchantApplicationIDLength,
      applicationIDData,
      _getAccountID(accountType),
      _getAccountLength(accountType, target),
      _formatAccount(accountType, target),
      countryID,
      countryLength,
      countryData,
      currencyID,
      currencyLength,
      bahtCurrencyData,
    ];

    if (amount != null) {
      data.add(amountID);
      data.add(_formatAmount(amount).length.toString().padLeft(2, '0'));
      data.add(_formatAmount(amount));
    }

    data.add(checksumID);
    data.add(checksumLength);

    var checksum = _getCrc16XMODEM()
        .convert(utf8.encode(data.join()))
        .toRadixString(16)
        .toUpperCase();

    return data.join() + checksum;
  }

  static String getPromptPayQRWithNewAmount(String qrData, double amount) {
    Iterable<PromptPayField> data =
        PromptPayData.fromQRData(qrData).asIterable();
    var amountData = amountID +
        _formatAmount(amount).length.toString().padLeft(2, '0') +
        _formatAmount(amount);
    var isAlreadyAddAmount = false;
    var newQRData = data.fold("", (qrData, element) {
      if (element != null && element.typeID == amountID) {
        isAlreadyAddAmount = true;
        return qrData + amountData;
      }

      if (element != null) {
        return qrData +
            element.typeID +
            element.length.toString().padLeft(2, '0') +
            element.data;
      }

      return qrData;
    });

    if (!isAlreadyAddAmount) {
      newQRData = newQRData + amountData;
    }

    newQRData = newQRData + checksumID + checksumLength;

    return newQRData +
        _getCrc16XMODEM()
            .convert(utf8.encode(newQRData))
            .toRadixString(16)
            .toUpperCase();
  }

  static Option<String> getAccountNumberFromQRData(String qrData) {
    var promptPayData = PromptPayData.fromQRData(qrData);

    if (promptPayData.transferring != null) {
      if (promptPayData.transferringPhoneNumber != null) {
        return Option.of(promptPayData.transferringPhoneNumber.data);
      }

      if (promptPayData.transferringIdentityNumber != null) {
        return Option.of(promptPayData.transferringIdentityNumber.data);
      }

      if (promptPayData.transferringEWallet != null) {
        return Option.of(promptPayData.transferringEWallet.data);
      }
    }

    if (promptPayData.billing != null) {
      return Option.of(promptPayData.billingID.data);
    }

    return Option.empty();
  }

  static bool isQRDataValid(String qrData) {
    if (qrData.length < 8) {
      return false;
    }

    final qrDataWithOutChecksum = qrData.substring(0, qrData.length - 4);
    final checksum = qrData.substring(qrData.length - 4, qrData.length);
    final newChecksum = _getCrc16XMODEM()
        .convert(utf8.encode(qrDataWithOutChecksum))
        .toRadixString(16)
        .toUpperCase();

    return newChecksum == checksum;
  }

  static String _getAccountID(AccountType accountType) {
    switch (accountType) {
      case AccountType.eWallet:
        return subMerchantAccountEWalletID;
      case AccountType.identityNumber:
        return subMerchantAccountIdentityID;
      default:
        return subMerchantAccountPhoneID;
    }
  }

  static String _getAccountLength(AccountType accountType, String target) {
    switch (accountType) {
      case AccountType.eWallet:
        return target.length.toString();
      case AccountType.identityNumber:
        return target.length.toString();
      default:
        return ("0066" + target.substring(1, target.length)).length.toString();
    }
  }

  static String _formatAccount(AccountType accountType, String target) {
    switch (accountType) {
      case AccountType.eWallet:
        return target;
      case AccountType.identityNumber:
        return target;
      default:
        return "0066" + target.substring(1, target.length);
    }
  }

  static String _formatAmount(double amount) {
    return amount.toStringAsFixed(2);
  }

  static ParametricCrc _getCrc16XMODEM() {
    // width=16 poly=0x1021 init=0x0000 refin=false refout=false xorout=0x0000 check=0x31c3 residue=0x0000 name="CRC-16/XMODEM"
    return new ParametricCrc(16, 0x1021, 0xFFFF, 0x0000,
        inputReflected: false, outputReflected: false);
  }
}
